V4L2 HEVC driver

6by9 opened the pull request:
This is @jc-kynesim 's driver with a load of checkpatch cleanups done. It's not quite checkpatch clean, but nearly.
- Some long lines (mainly as the V4L2 control defines are so long to start with)
- Uncommented spinlocks
- Uncommented memory barrier (possibly redundant as I've switched to readl/writel instead of John's volatile mapped accesses)
- No signed off on the main patch.
- MAINTAINERS - need to decide who to add to that.
- DT binding doc.

I've checked that it works with FFmpeg in the previous test cases.
John is still working through an issue with @popcornmix that Kodi stops displaying or something. Flagged as WIP until that is resolved.

ribalda committed:
media: add V4L2_CTRL_TYPE_AREA control type

Commit d1dc49370f8371b00e682ac409aa1987ce641e93 upstream.

This type contains the width and the height of a rectangular area.

Reviewed-by: Jacopo Mondi <jacopo@jmondi.org>
Signed-off-by: Ricardo Ribalda Delgado <ribalda@kernel.org>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

ribalda committed:
media: Documentation: media: Document V4L2_CTRL_TYPE_AREA

Commit 8ae3a0862993c09a8ef0f9abb379553370c517e3 upstream.

A struct v4l2_area containing the width and the height of a rectangular
area.

Reviewed-by: Jacopo Mondi <jacopo@jmondi.org>
Reviewed-by: Philipp Zabel <p.zabel@pengutronix.de>
Signed-off-by: Ricardo Ribalda Delgado <ribalda@kernel.org>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

ribalda committed:
media: add V4L2_CID_UNIT_CELL_SIZE control

Commit 61fd036d01111679b01e4b92e6bd0cdd33809aea upstream.

This control returns the unit cell size in nanometres. The struct provides
the width and the height in separated fields to take into consideration
asymmetric pixels and/or hardware binning.
This control is required for automatic calibration of sensors/cameras.

Reviewed-by: Philipp Zabel <p.zabel@pengutronix.de>
Signed-off-by: Ricardo Ribalda Delgado <ribalda@kernel.org>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

bparrot committed:
media: v4l2-common: add pixel encoding support

Commit d5a897c8428b38053df4b427a4277b1a0722bfa0 upstream.

It is often useful to figure out if a pixel_format is either YUV or RGB
especially for driver who can perform the pixel encoding conversion.

Instead of having each driver implement its own "is_this_yuv/rgb"
function based on a restricted set of pixel value, it is better to do
this in centralized manner.

We therefore add a pixel_enc member to the v4l2_format_info structure to
quickly identify the related pixel encoding.
And add helper functions to check pixel encoding.

Signed-off-by: Benoit Parrot <bparrot@ti.com>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

bparrot committed:
media: v4l2-common: add RGB565 and RGB55 to v4l2_format_info

Commit b373f84d77e1c409aacb4ff5bb5726c45fc8b166 upstream.

Add RGB565 and RGB555 to the v4l2_format_info table.

Signed-off-by: Benoit Parrot <bparrot@ti.com>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

Commit:
media: vb2: add V4L2_BUF_FLAG_M2M_HOLD_CAPTURE_BUF

This patch adds support for the V4L2_BUF_FLAG_M2M_HOLD_CAPTURE_BUF
flag.

It also adds a new V4L2_BUF_CAP_SUPPORTS_M2M_HOLD_CAPTURE_BUF
capability.

Drivers should set vb2_queue->subsystem_flags to
VB2_V4L2_FL_SUPPORTS_M2M_HOLD_CAPTURE_BUF to indicate support
for this flag.

Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

Commit:
media: v4l2-mem2mem: support held capture buffers

Commit f8cca8c97a63d77f48334cde81d15014f43530ef upstream.

Check for held buffers that are ready to be returned to vb2 in
__v4l2_m2m_try_queue(). This avoids drivers having to handle this
case.

Add v4l2_m2m_buf_done_and_job_finish() to correctly return source
and destination buffers and mark the job as finished while taking
a held destination buffer into account (i.e. that buffer won't be
returned). This has to be done while job_spinlock is held to avoid
race conditions.

Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

Commit:
media: videodev2.h: add V4L2_DEC_CMD_FLUSH

Add this new V4L2_DEC_CMD_FLUSH decoder command and document it.

Reviewed-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Alexandre Courbot <acourbot@chromium.org>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

jernejsk committed:
media: v4l2-mem2mem: add stateless_(try_)decoder_cmd ioctl helpers

Commit bef41d93aac64b54c3008ca6170bec54f85784f5 upstream.

These helpers are used by stateless codecs when they support multiple
slices per frame and hold capture buffer flag is set. It's expected that
all such codecs will use this code.

Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
Co-developed-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

Commit:
media: v4l2-mem2mem: add new_frame detection

Commit f07602ac388723233e9e3c5a05b54baf34e0a3e9 upstream.

Drivers that support VB2_V4L2_FL_SUPPORTS_M2M_HOLD_CAPTURE_BUF
typically want to know if a new frame is started (i.e. the first
slice is about to be processed). Add a new_frame bool to v4l2_m2m_ctx
and set it accordingly.

Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

paulkocialkowski committed:
media: v4l: Add definitions for HEVC stateless decoding

This introduces the required definitions for HEVC decoding support with
stateless VPUs. The controls associated to the HEVC slice format provide
the required meta-data for decoding slices extracted from the bitstream.

They are not exported to the public V4L2 API since reworking this API
will likely be needed for covering various use-cases and new hardware.

Multi-slice decoding is exposed as a valid decoding mode to match current
H.264 support but it is not yet implemented.

The interface comes with the following limitations:
* No custom quantization matrices (scaling lists);
* Support for a single temporal layer only;
* No slice entry point offsets support;
* No conformance window support;
* No VUI parameters support;
* No support for SPS extensions: range, multilayer, 3d, scc, 4 bits;
* No support for PPS extensions: range, multilayer, 3d, scc, 4 bits.

Signed-off-by: Paul Kocialkowski <paul.kocialkowski@bootlin.com>
[hverkuil-cisco@xs4all.nl: use 1ULL in flags defines in hevc-ctrls.h]
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

paulkocialkowski committed:
media: pixfmt: Document the HEVC slice pixel format

Commit de06f289283298e2938445019999cec46435375c upstream.

Document the current state of the HEVC slice pixel format.
The format will need to evolve in the future, which is why it is
not part of the public API.

Signed-off-by: Paul Kocialkowski <paul.kocialkowski@bootlin.com>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

jernejsk committed:
media: v4l2-mem2mem: Fix hold buf flag checks

Commit 1076df3a77b490d33429560a9e0603b3673223e2 upstream.

Hold buf flag is set on output queue, not capture. Fix that.

Fixes: f07602ac3887 ("media: v4l2-mem2mem: add new_frame detection")
Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Signed-off-by: Mauro Carvalho Chehab <mchehab@kernel.org>

jernejsk committed:
media: uapi: hevc: Add scaling matrix control

Taken from https://patchwork.linuxtv.org/patch/60728/
Changes (mainly documentation) have been requested.

HEVC has a scaling matrix concept. Add support for it.

Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>

jernejsk committed:
media: uapi: hevc: Add segment address field

From https://patchwork.linuxtv.org/patch/60725/
Changes requested, but mainly docs.

If HEVC frame consists of multiple slices, segment address has to be
known in order to properly decode it.

Add segment address field to slice parameters.

Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>

6by9 committed:
media: videodev2.h: Add a format for column YUV4:2:0 modes

Some of the Broadcom codec blocks use a column based YUV4:2:0 image
format, so add the documentation and defines for both 8 and 10 bit
versions.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

jc-kynesim committed:
media: v4l2-mem2mem: allow request job buffer processing after job finish

Allow the capture buffer to be detached from a v4l2 request job such
that another job can start before the capture buffer is returned. This
allows h/w codecs that can process multiple requests at the same time
to operate more efficiently.

Signed-off-by: John Cox <jc@kynesim.co.uk>

jc-kynesim committed:
staging: media: Add Raspberry Pi V4L2 H265 decoder

This driver is for the HEVC/H265 decoder block on the Raspberry
Pi 4, and conforms to the V4L2 stateless decoder API.

Signed-off-by: John Cox <jc@kynesim.co.uk>

6by9 committed:
dtoverlays: Add overlay to enable the HEVC V4L2 driver

This replaces the rpivid_mem register mapping driver.
When the driver is complete, these DT changes should be
merged into the base DT instead of being an overlay.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
defconfig: Add V4L2 HEVC driver to bcm2711_defconfig

This replaces the register mapped driver.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

popcornmix commented:
I think there are a couple of additional commits (either on my kodi/rpi4-kms tree or on buildbot).
"Stash: add threaded callback" "Dynamically size all buffers" which fixes an issue when hardware signals buffer sizes are insufficient and needs a realloc/retry.

I think the current belief is the kodi black screen issue only occurs with the hardest streams and is a bad interaction when running behind. It may be something that can be improved, but otherwise isn't a blocker.

6by9 commented on drivers/media/common/v4l2-tpg/v4l2-tpg-core.c:
```diff
@@ -309,6 +309,7 @@ bool tpg_s_fourcc(struct tpg_data *tpg, u32 fourcc)
 	case V4L2_PIX_FMT_NV21M:
 		tpg->buffers = 2;
 		/* fall through */
+	case V4L2_PIX_FMT_SAND8:
```
Do we actually need to use the test pattern generator?

jc-kynesim replied:
Probably not - I think I added that line when I was adding SAND cases everywhere I could find NV12 used to make sure I got frames through

6by9 commented on Documentation/media/uapi/v4l/pixfmt-nv12.rst:
```diff
@@ -38,6 +38,14 @@ with a Cr byte.
+requires knowledge of the stride between columns, therefore the bytesperline
+value has been repurposed to denote the number of 128 byte long lines between
+the start of each column.
+
```
Docs need finishing, and 10bit version adding.

6by9 commented on include/uapi/linux/videodev2.h:
```diff
@@ -738,6 +738,8 @@ struct v4l2_pix_format {
 #define V4L2_PIX_FMT_INZI     v4l2_fourcc('I', 'N', 'Z', 'I') /* Intel Planar Greyscale 10-bit and Depth 16-bit */
 #define V4L2_PIX_FMT_SUNXI_TILED_NV12 v4l2_fourcc('S', 'T', '1', '2') /* Sunxi Tiled NV12 Format */
 #define V4L2_PIX_FMT_CNF4     v4l2_fourcc('C', 'N', 'F', '4') /* Intel 4-bit packed depth confidence information */
+#define V4L2_PIX_FMT_SAND8    v4l2_fourcc('S', 'A', 'N', 'D') /* RPI 8 bit sand */
```
What was wrong with the V4L2_PIX_FMT_NV12_COL128 format I'd defined in the earlier patch?

jc-kynesim replied:
> What was wrong with the V4L2_PIX_FMT_NV12_COL128 format I'd defined in the earlier patch?

Probably nothing - either I wrote that line before I found your patch or I failed to notice it.

6by9 commented on drivers/media/v4l2-core/v4l2-ctrls.c:
```diff
@@ -974,6 +974,7 @@ const char *v4l2_ctrl_get_name(u32 id)
 	case V4L2_CID_MPEG_VIDEO_HEVC_SPS:			return "HEVC Sequence Parameter Set";
 	case V4L2_CID_MPEG_VIDEO_HEVC_PPS:			return "HEVC Picture Parameter Set";
 	case V4L2_CID_MPEG_VIDEO_HEVC_SLICE_PARAMS:		return "HEVC Slice Parameters";
+	case V4L2_CID_MPEG_VIDEO_HEVC_SCALING_MATRIX:	return "HEVC Scaling Matrix";
```
Need to expand out the commit text, and add a Signed-off-by:

6by9 reviewed (commented):
Sorry, commenting on my own PR as things are more obvious as soon as you've created it.
Still a couple more jobs to do.

6by9 commented:
John had told me 91187ff10c0447762b641cc7757fadf6914b3696 on his branch test/5.4/v4l2_5 of buildbot was the latest. I'm happy to squash other commits in, but wanted/needed to get it checkpatch clean before considering submitting.
John was happy that redoing subsequent patches on top of a cleaned up base shouldn't be too onerous.

popcornmix commented:
Sure. Should be functional (for vast majority of streams) without the additional patches.
No problem with getting this merged soon and adding to it later.

jc-kynesim commented:
> I think there are a couple of additional commits (either on my kodi/rpi4-kms tree or on buildbot).
> "Stash: add threaded callback" "Dynamically size all buffers" which fixes an issue when hardware signals buffer sizes are insufficient and needs a realloc/retry.
> I think the current belief is the kodi black screen issue only occurs with the hardest streams and is a bad interaction when running behind. It may be something that can be improved, but otherwise isn't a blocker.

So is your current belief that my dev/5.4/fix_pu_alloc_1 branch actually works? 'cos if so then I would like to merge that in now

6by9 commented:
Tell me which patches are required and I'll merge them in here, along with sorting out the docs and other tasks. Ideally we really do want rpivid all as one squashed commit if we're looking to upstream. 

(Need to get a mainline 5.6 tree booting on a Pi4 before upstreaming to be able to test it).

popcornmix commented:
> So is your current belief that my dev/5.4/fix_pu_alloc_1 branch actually works? 'cos if so then I would like to merge that in now

It clearly fixes an issue when buffer required by hardware exceeds default size. 
ffmpeg can now decode the whole file rather than stalling at the big frame.

It allows the problematic file to get further in kodi but we see other issues caused by performance being on the edge.

Overall it seems beneficial so happy to include it.

jc-kynesim commented:
> Uncommented spinlocks
What sort of comments are wanted?  - I'm happy to add them

> Uncommented memory barrier (possibly redundant as I've switched to readl/writel instead of John's volatile mapped accesses)
Not actually the best fix I think - it adds many more memory barrier ops than needed - the bulk of them just want to be writel_relaxed() with only the last in the set being a writel()

jc-kynesim commented:
> Tell me which patches are required and I'll merge them in here, along with sorting out the docs and other tasks. Ideally we really do want rpivid all as one squashed commit if we're looking to upstream.
> (Need to get a mainline 5.6 tree booting on a Pi4 before upstreaming to be able to test it).

As the branch is currently set up for debug it is probably easiest (though non-trivial) for me to recreate it on your patch and remove the unwanted debug as I go rather than get you to do it - I'll also fix up spinlock comments etc.

6by9 commented:
>> Uncommented spinlocks

> What sort of comments are wanted? - I'm happy to add them

What's being protected by them. Either explicitly commented, or blocked together, eg
https://elixir.bootlin.com/linux/latest/source/include/media/v4l2-dev.h#L289
```
	/* V4L2 file handles */
	spinlock_t		fh_lock;
	struct list_head	fh_list;

```
If you provide guidance, I can make the edits.
>> Uncommented memory barrier (possibly redundant as I've switched to readl/writel instead of John's volatile mapped accesses)

> Not actually the best fix I think - it adds many more memory barrier ops than needed - the bulk of them just want to be writel_relaxed() with only the last in the set being a writel()

If so then we can make a new function the final [write|read]l, but raw accesses to volatile mapped registers isn't going to be acceptable upstream.

lategoodbye commented on drivers/staging/media/rpivid/Kconfig:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: GPL-2.0
+
+# SPDX-License-Identifier: GPL-2.0
```
I think the second one isn't necessary

6by9 replied:
Thanks for the review. I thought this was worth saying twice ;-)

lategoodbye commented on drivers/staging/media/rpivid/Kconfig:
```diff
@@ -0,0 +1,17 @@
+	  Support for the Rpi H265 h/w decoder.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called sunxi-cedrus.
```
Oops, copy & paste

6by9 replied:
I'm also renaming the module to rpivid-hevc. rpivid-v4l2 has the potential to get confused against another codec block.

lategoodbye commented on drivers/staging/media/rpivid/rpivid_hw.c:
```diff
@@ -0,0 +1,259 @@
+	ret = devm_request_irq(dev->dev, irq_dec, rpivid_irq,
+			       0, dev_name(dev->dev), dev);
+	if (ret) {
+		dev_err(dev->dev, "Failed to request IRQ\n");
```
The return code might be interesting here

lategoodbye commented on drivers/staging/media/rpivid/rpivid_video.c:
```diff
@@ -0,0 +1,572 @@
+	return container_of(file->private_data, struct rpivid_ctx, fh);
+}
+
+/* constrian x to y,y*2 */
```
constrain?

lategoodbye commented on drivers/staging/media/rpivid/rpivid_video.c:
```diff
@@ -0,0 +1,572 @@
+/* constrian x to y,y*2 */
+static inline unsigned int constrain2x(unsigned int x, unsigned int y)
+{
+	return x < y ? y : x > y * 2 ? y : x;
```
this is very hard to read

popcornmix replied:
Based on comment I'd have expected ``x < y ? y : x > y * 2 ? y * 2 : x;``

6by9 replied:
I'll break it up onto multiple lines once someone confirms whether it is functionally correct or not.

jc-kynesim replied:
The original code was, in fact, correct - the comment was misleading.  The intention was to constrain buffer sizes - if they were in the range min to min*2 then accept them, otherwise choose min as if I'm going to reject the users suggestion I might as well choose the "correct" size.

lategoodbye commented on drivers/staging/media/rpivid/rpivid_video.c:
```diff
@@ -0,0 +1,572 @@
+				(ctb_log2_size_y - min_tb_log2_size_y) ||
+	    /* Check pcm stuff */
+	    sps->num_short_term_ref_pic_sets > 64 ||
+	    sps->num_long_term_ref_pics_sps > 32)
```
Please check for single error cases and return with 0

6by9 replied:
Where checking eg sps->pic_width_in_luma_samples is between 32 & 4096 I've kept it as one ```if```, but split the rest up.

lategoodbye commented on drivers/staging/media/rpivid/rpivid_video.c:
```diff
@@ -0,0 +1,572 @@
+	return 0;
+}
+
+static inline void cpy_colour(struct v4l2_pix_format *d,
```
copy_color ?

lategoodbye commented on drivers/staging/media/rpivid/rpivid_video.c:
```diff
@@ -0,0 +1,572 @@
+	.vidioc_streamoff		= v4l2_m2m_ioctl_streamoff,
+
+	.vidioc_try_decoder_cmd		=
+				v4l2_m2m_ioctl_stateless_try_decoder_cmd,
```
In this case it's okay to violate the 80 char limit

lategoodbye commented on drivers/staging/media/rpivid/rpivid_video.c:
```diff
@@ -0,0 +1,572 @@
+{
+	const struct v4l2_ctrl_hevc_sps * const sps =
+		rpivid_find_control_data(ctx, V4L2_CID_MPEG_VIDEO_HEVC_SPS);
+	struct v4l2_pix_format pix_fmt = {
```
According to the definition rpivid_find_control_data could return NULL, so this impossible in this and some other cases?

lategoodbye commented on drivers/staging/media/rpivid/rpivid.c:
```diff
@@ -0,0 +1,430 @@
+	.remove		= rpivid_remove,
+	.driver		= {
+		.name		= RPIVID_NAME,
+		.owner = THIS_MODULE,
```
Please drop this, it's already done by platform_driver

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+
+//////////////////////////////////////////////////////////////////////////////
+
+struct RPI_CMD {
```
i think this should be lower case

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+#define RPI_MVSTRIDE 0x8034
+#define RPI_COLBASE 0x8038
+#define RPI_COLSTRIDE 0x803C
+#define RPI_CURRPOC 0x8040
```
use tabs for indenting would improve readability

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+	int i;
+
+	for (i = 1; ctb >= bd[i]; i++)
+		; // bd[] has num+1 elements; bd[0]=0; see hevc_ps.c
```
i assume this source reference doesn't apply anymore

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+		136, 153, 154, 166, 183, 140, 136, 153, 154, 170, 153, 138, 138,
+		122, 121, 122, 121, 167, 151, 183, 140, 151, 183, 140, 0,   0,
+	},
+};
```
a comment about these values would be nice

6by9 replied:
@jc-kynesim Please advise as to what these are.

jc-kynesim replied:
CABAC probability init values - commented usefully in my current patch

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+	},
+};
+
+static void WriteProb(struct rpivid_dec_env *const de,
```
Please avoid camel case

6by9 replied:
I thought I'd killed it all as checkpatch had complained in places :-( Done

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+		dst0[0] = dc;
+		break;
+	}
+	}
```
the indentation looks broken

6by9 replied:
It's not actually as it's the end of the case which was declaring variables in that scope, and the end of the overall switch statement, but I'll agree it looks strange.

Seeing as d, x, and y were used in 2 of the 4 cases I've switched to defining them at the top of the function.

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+	//              run->src->timecode.seconds,
+	//              run->src->timecode.frames);
+	//    v4l2_info(&dev->v4l2_dev, "vptr=%p\n",
+	//    vb2_plane_vaddr(&run->src->vb2_buf, 0));
```
Please drop this or make it a debug message

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+	unsigned int cmd_len, cmd_max;
+	unsigned int num_slice_msgs;
+	unsigned int PicWidthInCtbsY;
+	unsigned int PicHeightInCtbsY;
```
please no camel case

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
i think it's preferred to use kzalloc directly

6by9 replied:
I'm not sure if it's what you mean, but the function is actually giving us no gain.
```if (!s) return NULL``` is equivalent to ```return s```, and that is what happens in the success case. I'll drop the function.

lategoodbye replied:
Yes, this is what i tried to say

lategoodbye commented on drivers/staging/media/rpivid/rpivid_h265.c:
```diff
@@ -0,0 +1,2075 @@
+
+// Returns:
+// -2 Other error
+// -1 Out of coeff space
```
Maybe we can use standard error code like -ENOMEM

jc-kynesim replied:
Internal status not communicated back to user and anyway rewritten in my current patch

6by9 replied:
The caller does nothing other than ```if (check_status != 0)```, so I've amended it to -EINVAL and -ENOMEM.

6by9 commented:
Branch updated based on lategoodbye's review comments. Now to fix up my own.

6by9 committed:
media: dt-bindings: media: Add binding for the Raspberry PI HEVC decoder

Adds a binding for the HEVC decoder found on the BCM2711 / Raspberry Pi 4.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Branch https://github.com/6by9/linux/tree/rpi-5.4.y-hevc-wip pushed with fixups, early DT binding, and John's latest improvements. Now going to squash it down to update this PR.
(Still need to complete the V4L2 pixel format docs)

Note that https://github.com/6by9/linux/commit/1c73ca4e799f03217cea879eaf40d6c601a2b0e6 needs to be dropped. SAND as a fourcc means nothing to anyone outside Broadcom/Pi Towers, but it needs FFmpeg to be updated to match, and I haven't rebuilt that at present.

6by9 commented on drivers/staging/media/rpivid/rpivid_video.c:
```diff
@@ -0,0 +1,588 @@
+
+	switch (ctx->src_fmt.pixelformat) {
+	case V4L2_PIX_FMT_HEVC_SLICE:
+		ctx->current_codec = RPIVID_CODEC_H265;
```
@jc-kynesim This only ever gets set to RPIVID_CODEC_H265.

Based on other recent conversations, do we really see this driver supporting multiple codecs? I read the conclusion as "no", therefore we can strip out a load of indirection.

jc-kynesim replied:
I've stripped some out from the original but not all so go ahead - though you probably want to keep the checks that we are, in fact, being asked to do what we expect.

6by9 replied:
I have kept the check that we are getting V4L2_PIX_FMT_HEVC_SLICE, but anything else should be impossible to get through rpivid_s_fmt_vid_cap anyway.

I have also kept dev->dec_ops as a function pointer, so actually you could change the function pointer on change of format to achieve the same end result.

popcornmix commented:
Any reason not to add this to arm64 defconfig?

6by9 commented:
> Any reason not to add this to arm64 defconfig?

Probably not, but I haven't tried it.
Allocations are from the CMA heap, so those should still be under the same restrictions as 32-but world.

lategoodbye commented:
> Branch https://github.com/6by9/linux/tree/rpi-5.4.y-hevc-wip pushed with fixups, early DT binding, and John's latest improvements. Now going to squash it down to update this PR.
> (Still need to complete the V4L2 pixel format docs)

New DT binding should be in YAML format:
https://www.kernel.org/doc/Documentation/devicetree/writing-schema.md

Unfortunately i don't know a good tutorial.

Finally you need to make sure that this step succeed:
make dtbs_check DT_SCHEMA_FILES=Documentation/devicetree/bindings/media/rpivid_hevc.yaml

6by9 commented:
> New DT binding should be in YAML format:

Thank you. As I was backporting to 5.4 they are all still text files. Yes, copy/paste is always a good starting point.

6by9 commented:
Docs completed and look reasonable.
DT binding updated to YAML.

We have 3 patches from @jc-kynesim without Signed-Off-By: lines.
John - please review the patches and confirm if you're happy for me to attach them. Refer to https://www.kernel.org/doc/html/latest/process/submitting-patches.html#developer-s-certificate-of-origin-1-1 as to the meaning of SoB.

The one bit that may be wanted is clock control. The firmware appears to have clock control for the HEVC block, but this driver isn't controlling it. Once Maxime's updated clock driver is merged then it becomes easy, so I'm tempted to skip it for now. It can be added later as an optional clock property,
Thoughts?

popcornmix commented:
Clock control is needed (I believe it was in @jc-kynesim 's earlier ffmpeg hevc driver).
It should read the max hevc clock and request that while it's active.
Otherwise we'll run at half speed when the arm is <50% busy (which it typically is) and fail to keep up.

6by9 commented:
As above, clock control is going to need to wait for Maxime's driver to be merged, or at least the discussions completed.
I don't fancy adding it into clk-bcm2835 unless really needed.

popcornmix commented:
It can only safely use the firmware clock driver (or mailbox directly, but that would be frowned upon upstream).

We can merge this (with note that performance isn't full without force_turbo=1 or performance governor) and add the clock request when available.

6by9 commented:
```vcgencmd measure_clock hevc``` is reporting 250MHz for me on an idle system. I can't find the max permitted quickly.

It looks like there are ```hevc_freq_min``` and ```hevc_freq_max``` config.txt options, so that would be one way around it.
Memory says there is a way of allocating a NULL clock (code review saying NULL was a valid return from ```devm_clk_get```). If I can quickly find the relevant bit for that then I'll add it to the current configuration, and we can replace it with the firmware clock driver at a suitable point in the future.

Do we just always ask for max, or is there any better logic around?

popcornmix commented:
This was logic for ffmpeg based hevc driver https://github.com/popcornmix/FFmpeg/commit/d9d273b9a642565807bf63e55457312cccd23e37

jc-kynesim commented:
> We have 3 patches from @jc-kynesim without Signed-Off-By: lines.
> John - please review the patches and confirm if you're happy for me to attach them. Refer to https://www.kernel.org/doc/html/latest/process/submitting-patches.html#developer-s-certificate-of-origin-1-1 as to the meaning of SoB.

I'm certainly happy with the decoder patch, at one point I'm pretty sure the mem2mem patch had a signed off line when seems to have become lost so I'm happy with that, the "scaling matrix" patch is mostly jernejs work (I've added a flag) & has, at least in part I think, been submitted to upstream once where it faded into oblivion.

popcornmix commented:
@6by9 needs a rebase after 5.4 update.
What else does this need before merging?

6by9 commented:
> What else does this need before merging?

Clock driver hooks.
I have what I hope is the correct code, and DT is advertising a fixed clock for now. 
I'm trying to get it tested and pushed today, but also need to head into the office to collect some kit.

6by9 committed:
media: hevc_ctrls: Add slice param dependent slice segment

Adds V4L2_HEVC_SLICE_PARAMS_FLAG_DEPENDENT_SLICE_SEGMENT define.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Updated and with clock changes (fixed clock defined in DT overlay at present).

@jc-kynesim I've added your SoB to the driver, cherry-picked jernej's patch from https://patchwork.linuxtv.org/patch/60728/ and added an extra one that adds your one define, and hopefully now have a suitable tree!

6by9 commented:
Except I've messed something up in the last fixup! I've obviously lost part of John's patch as slice_segment_addr and num_entry_point_offsets are missing.

popcornmix commented:
What effect does the 500MHz hevc clock request in DT have?

6by9 commented:
@jc-kynesim So you appear to have merged 
- https://patchwork.linuxtv.org/patch/60728/
- https://patchwork.linuxtv.org/patch/60725/
- Adding define V4L2_HEVC_SLICE_PARAMS_FLAG_DEPENDENT_SLICE_SEGMENT
- And added num_entry_point_offsets and entry_point_offset_minus1[256] from somewhere. Do you have any idea as where?

With these being uapi headers that clients will be using, heritage is extremely important.

6by9 commented:
> What effect does the 500MHz hevc clock request in DT have?

Absolutely none.
Until we have Maxime's clock driver this is just a place holder so that the driver can ask a clock to run at 500MHz without error. There is currently no clock control happening at all.

pelwell commented:
Yes - it's just a way to pass a clock frequency into the driver.

popcornmix commented:
Okay just checking it didn't magically drive the upstream clock driver.

Ideally the 500MHz shouldn't be hard coded, but queried from firmware (so it respects changes in firmware clocking or config.txt ``hevc_clock`` settings).
I believe the firmware clock driver does squirrel away the max/min of each clock queried from firmware.
But that can be discussed later. Would be good to get this merged soon.

6by9 committed:
media: uapi: Add hevc ctrls for WPP decoding

WPP can allow greater parallelism within the decode, but needs
offset information to be passed in.

Adds num_entry_point_offsets and entry_point_offset_minus1 to
v4l2_ctrl_hevc_slice_params.

This is based on Jernej Skrabec's patches for cedrus which
implement the same feature.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
https://github.com/raspberrypi/linux/pull/3505/commits/5d7d08109a2c6972ae1e68cf40915f66548ef1a3 needs explanation as to where it's come from, otherwise I think this is good to go.

Clock stuff needs to be revisited, but there's no point fussing over it until we have a working fw clock driver merged. At present it doesn't query the frequency range on general clocks, and I haven't found a suitable API for extracting that from a clock anyway.

jc-kynesim commented:
> @jc-kynesim So you appear to have merged
> 
> * https://patchwork.linuxtv.org/patch/60728/
> * https://patchwork.linuxtv.org/patch/60725/
> * Adding define V4L2_HEVC_SLICE_PARAMS_FLAG_DEPENDENT_SLICE_SEGMENT
> * And added num_entry_point_offsets and entry_point_offset_minus1[256] from somewhere. Do you have any idea as where?
> 
> With these being uapi headers that clients will be using, heritage is extremely important.

Apart from V4L2_HEVC_SLICE_PARAMS_FLAG_DEPENDENT_SLICE_SEGMENT I got all of that by simply copying jernejs (at least I think its his) hevc-ctrl.c from the current Kodi tree.

6by9 commented:
> I got all of that by simply copying jernejs (at least I think its his) hevc-ctrl.h from the current Kodi tree.

Being a pedant, Kodi is a userspace app and this is a kernel header. Checking https://github.com/xbmc/xbmc there is no file hevc-ctrls.h in the tree.

Please can you provide a link to the tree that you used so I can cherry-pick the commit.

jc-kynesim commented:
> > I got all of that by simply copying jernejs (at least I think its his) hevc-ctrl.h from the current Kodi tree.
> 
> Being a pedant, Kodi is a userspace app and this is a kernel header. Checking https://github.com/xbmc/xbmc there is no file hevc-ctrls.h in the tree.
> 
> Please can you provide a link to the tree that you used so I can cherry-pick the commit.

Mea culpa - I misremembered - it came from the LibreELEC distribution - specifically projects/Allwinner/patches/linux/0005-cedrus-improvements.patch gives num_entry_point_offsets and entry_point_offset_minus1 - which came from  https://github.com/lrusak/LibreELEC.tv.git branch ffmpeg-update - commit 203d54d9e8f376a57b643fe3de19d4ff70a35c3c

Sorry about that

jc-kynesim commented:
>> I got all of that by simply copying jernejs (at least I think its his) hevc-ctrl.h from the current Kodi tree.
>
>Being a pedant, Kodi is a userspace app and this is a kernel header. Checking https://github.com/xbmc/xbmc there is no file hevc-ctrls.h in the tree.
>
>Please can you provide a link to the tree that you used so I can cherry-pick the commit.

OK - Sorry about the confusion here - in particular depending on how
github works you may have seen that I created a comment which I then
deleted ('cos I thought I was talking rubbish) which I now find was
correct.  I'll try harder this time.

When I said Kodi I actually meant LibreELEC.tv

The patch for entry_point_offset_minus1 cames from

repo: https://github.com/lrusak/LibreELEC.tv.git
branch: ffmpeg-update
commit: 203d54d9e8f376a57b643fe3de19d4ff70a35c3c
file: /projects/Allwinner/patches/linux/0005-cedrus-improvements.patch

but I will have applied all the paches in that dir - many will not
matter.

Regards

JC

6by9 commented:
So it looks like you only use num_entry_point_offsets and not the entry_point_offset_minus1 array in wpp_decode_slice. How does that work?
I've added logging for num_entry_point_offsets and it doesn't seem to get triggered. Is this some specific optional extension that only some clips will use?

jc-kynesim commented:
>So it looks like you only use num_entry_point_offsets and not the entry_point_offset_minus1 array in wpp_decode_slice. How does that work?
>I've added logging for num_entry_point_offsets and it doesn't seem to get triggered. Is this some specific optional extension that only some clips will use?

Yes - WPP allows you to process the "MB" (not called that anymore in
H.265) lines of a slice in parallel - the info (and appropriate
encoding) will only be there for some clips (it is quite common but by
no means universal).  You only really need the offsets if you are
actually going to do stuff in parallel and we don't.

Does that help?  I didn't add the offset array but it might clearly be
needed - I guess the Cedrus codec may need them.

jc-kynesim commented:
And as a FYI to my previous comment - the Pi3 H265 decode does use the offsets to parallelise decode and it does help.

6by9 commented:
Thanks. I've pinged jernej on Slack asking if there is a plan to upstream it.
If not then I'll just amend commit text of my current patch and we'll run with it for now. I can foresee pushback again because num_entry_point_offsets has been dropped into the middle of the structure, which means a breakage for existing clients. It's not in the released uapi yet, so it doesn't really make any difference.

6by9 commented:
WIP tag removed.

I've reworded the commit text for the WPP properties patch, and reordered it so that all steps should compile (I haven't tested that).

If someone else could do a test and confirm that this is still functional, then I think it's good to go.
Please note that the V4L2 pixel formats have changed in both four cc and define. FFmpeg will need updating to match.

popcornmix commented:
@6by9 I updated ffmpeg: `V4L2_PIX_FMT_SAND8`->`V4L2_PIX_FMT_NV12_COL128` and `V4L2_PIX_FMT_SAND30`->`V4L2_PIX_FMT_NV12_10_COL128`
and I could play both 8-bit hevc and 10-bit through fkms with hw decode speed.

Also worked with kodi (with fkms)

So I think this is good to go.

6by9 commented:
@pelwell Any other requests from you?
Otherwise can we merge before popcornmix next rebases to a new upstream release?

pelwell commented:
Just a few minor overlay niggles:
1. The overlay is in the wrong position in the Makefile (off-by-one).
2. The compatible string should be "brcm,bcm2711" because it is strictly 2711-only (otherwise the test script will try to apply it to other platforms and fail).
3. It needs an entry in the README file.

6by9 commented:
I've made those 3 changes, so hopefully your overlay test scripts will now be happy.

pelwell commented:
I'm seeing a warning:
```
drivers/staging/media/rpivid/rpivid_h265.c: In function 'write_bitstream':
drivers/staging/media/rpivid/rpivid_h265.c:534:2: warning: #warning "EMU set - probably shouldn't be wanted" [-Wcpp]
 #warning "EMU set - probably shouldn't be wanted"
```

6by9 commented:
@jc-kynesim I meant to raise that warning with you before.
```
drivers/staging/media/rpivid/rpivid_h265.c: In function 'write_bitstream':
drivers/staging/media/rpivid/rpivid_h265.c:534:2: warning: #warning "EMU set - probably shouldn't be wanted" [-Wcpp]
 #warning "EMU set - probably shouldn't be wanted"
```

jc-kynesim commented:
> I'm seeing a warning:
> 
> ```
> drivers/staging/media/rpivid/rpivid_h265.c: In function 'write_bitstream':
> drivers/staging/media/rpivid/rpivid_h265.c:534:2: warning: #warning "EMU set - probably shouldn't be wanted" [-Wcpp]
>  #warning "EMU set - probably shouldn't be wanted"
> ```

Yeah - note from me to me - the current ffmpeg code passes bitstreams with emulation prevention left in. My reading of the "standard" suggests that it shouldn't. The warning could be removed - I left it there as a hard note that I thought I was doing something wrong.

I could also try and "fix" FFmpeg and remove the warning and h/w setup.  There are some tedious implications to doing so though.

pelwell commented:
If it's expected or unavoidable then I'd prefer it to be removed.

jc-kynesim commented:
My actual guess is that the current behaviour will become the standard for this stream type and another stream type may be invented if anyone ends up caring.  So it is probably safe / sensible to remove the warning - replace it with a comment.

6by9 commented:
> So it is probably safe / sensible to remove the warning - replace it with a comment.

OK, will do.

6by9 commented:
Done.

popcornmix commented:
@pelwell okay to merge?

hvenev commented:
> @6by9 I updated ffmpeg: `V4L2_PIX_FMT_SAND8`->`V4L2_PIX_FMT_NV12_COL128` and `V4L2_PIX_FMT_SAND30`->`V4L2_PIX_FMT_NV12_10_COL128`
> and I could play both 8-bit hevc and 10-bit through fkms with hw decode speed.
> 
> Also worked with kodi (with fkms)

Where can I find this version of ffmpeg?

6by9 commented:
@hvenev https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_1

hvenev commented:
> @hvenev https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_1

That seems to mention `V4L2_PIX_FMT_SAND8` and `V4L2_PIX_FMT_SAND30` but not `V4L2_PIX_FMT_NV12_COL128` and `V4L2_PIX_FMT_NV12_10_COL128`.

Is this patch more recent: https://github.com/popcornmix/xbmc/blob/gbm/tools/depends/target/ffmpeg/0001-rpi-Add-hevc-acceleration.patch ?

6by9 commented:
I took that branch and replaced ```V4L2_PIX_FMT_SAND8```->```V4L2_PIX_FMT_NV12_COL128``` and ```V4L2_PIX_FMT_SAND30```->```V4L2_PIX_FMT_NV12_10_COL128```.

I don't know what has been released/latest version - I leave that to @popcornmix and @jc-kynesim.

jc-kynesim commented:
The Kodi branch is the only one which currently has the V4L2 work in it.  Annoyingly it is a different base version to the standard Debian one.  I'm guessing Kodi must still be using an older header file to have the older #defines.  I guess that the Debian port should be updated to have the V4L2 work in it too now we have the possibility of using it.

popcornmix commented:
https://github.com/popcornmix/xbmc/blob/gbm/tools/depends/target/ffmpeg/0001-rpi-Add-hevc-acceleration.patch is the source being used by LibreELEC kodi builds and works with kernels including this PR.
The tree this came from is here: https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_4_2_2

6by9 commented:
> The tree this came from is here: https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_4_2_2

Can we get that merged into https://github.com/Rpi-distro/FFmpeg, or are we waiting on something else?

popcornmix commented:
It's complicated by kodi using ffmpeg 4.2.2 and debian buster using 4.1.4.

@jc-kynesim is https://github.com/jc-kynesim/rpi-ffmpeg/tree/test/4.1.4/rpi_main the right tree for debian? Can we get a version that supports 5.4 v4l2 and still works with existing VLC?

jc-kynesim commented:
Yes that should be the right tree.  I'll work on the port once I'm done with my current VAAPI work which may impact our FFmpeg port as well.  We will need to get rid of the horrid config kludge that enables/disables drm_prime passthrough too and replace it either with a command line switch if I can work out how to get one passed to where I need it or preferably some sort of automagic.

hvenev commented:
Please note that ffmpeg with that patch fails to build:
 - without `--enable-rpi` because various SAND methods are undefined (`CONFIG_SAND` is disabled).
 - with `--enable-rpi` because MMAL_ES_FORMAT_FLAG_COL_FMTS_WIDTH_IS_COL_STRIDE` is not defined (and that macro is not mentioned in https://github.com/raspberrypi/userland)

jc-kynesim commented:
MMAL_ES_FORMAT_FLAG_COL_FMTS_WIDTH_IS_COL_STRIDE is defined in /opt/vc/include/interface/mmal/mmal_format.h on my pi4 and I don't think that is 'cos I've edited it.

6by9 commented:
> MMAL_ES_FORMAT_FLAG_COL_FMTS_WIDTH_IS_COL_STRIDE is defined in /opt/vc/include/interface/mmal/mmal_format.h on my pi4 and I don't think that is 'cos I've edited it.

I suspect we've got another patch that hasn't been merged from firmware to userland :-(

hvenev commented:
It appears that they are only present in the repo with the 32-bit blobs of unknown origin:

 - https://github.com/raspberrypi/firmware/search?q=MMAL_ES_FORMAT_FLAG_COL_FMTS_WIDTH_IS_COL_STRIDE&type=
 - https://github.com/raspberrypi/userland/search?q=MMAL_ES_FORMAT_FLAG_COL_FMTS_WIDTH_IS_COL_STRIDE&type=

hvenev commented:
> > MMAL_ES_FORMAT_FLAG_COL_FMTS_WIDTH_IS_COL_STRIDE is defined in /opt/vc/include/interface/mmal/mmal_format.h on my pi4 and I don't think that is 'cos I've edited it.
> 
> I suspect we've got another patch that hasn't been merged from firmware to userland :-(

Can't you just rsync them or something?

6by9 commented:
> Can't you just rsync them or something?

Two git trees, userland being a subset of firmware. Applying with history isn't totally trivial, and someone needs to pick up on the fact that it's needed.
https://github.com/raspberrypi/userland/pull/624

hvenev commented:
Just out of curiosity, are there any other patches for which the source code isn't available in [raspberrypi/userland](https://github.com/raspberrypi/userland) but are a part of the 32-bit blobs of questionable origin at [raspberrypi/firmware](https://github.com/raspberrypi/firmware)?

hvenev commented:
Also note that even with the userland pull request, ffmpeg still fails to build with `--enable-rpi`:
```
libavcodec/rpi_mailbox.h:47:13: error: size of array 'vc_image_t_size_check' is negative
   47 | typedef int vc_image_t_size_check[(sizeof(VC_IMAGE_T) == 64) * 2 - 1];
      |             ^~~~~~~~~~~~~~~~~~~~~
```

hvenev commented:
This time I tried `--disable-rpi` and I fixed the SAND mess.

Another build of ffmpeg later:
```
/usr/bin/ld: libavutil/libavutil.so: undefined reference to `rpi_sand128b_stripe_to_8_10'
```

It looks like the function is used when `HAVE_NEON`, but it's really only implemented when `defined(__arm__) && HAVE_NEON`.

I will send a pull request when I get ffmpeg to build (and hopefully work).

jc-kynesim commented:
I don't know if it will help you but if you have https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_4_2_2 checked out then you will find a shell script pi-util/conf_native.sh which contains the set of config options I use for compiling that code on a pi4 (and it does compile - I've just checked).  The cross-compile scripts don't currently work on this codebase.

I normally spot the userland/firmware issues you are having as they show up in the cross compiles but whilst all the kernel & v4l2 stuff that are needed for V4L2 were still up in the air I found it too hard to keep sysroots in any sort of sync and resorted to compiling on the pi and so failed to spot  the fact that stuff was out of sync (I have no power to fix this up - only to winge).

Anyway - good luck

jc-kynesim commented:
> It looks like the function is used when `HAVE_NEON`, but it's really only implemented when `defined(__arm__) && HAVE_NEON`.

OK - I'll bite - how do you have neon and not arm?

popcornmix commented:
https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_4_2_2 is more recent and should match the kodi patch.

nkichukov commented:
I still fail to compile kodi-gbm with ffmpeg kodi_v4l2_4_2_2 both from popcornmix repo on rpi4 with 32-bit userland, the failure is:

```
src/libavcodec/v4l2_request.c:165:5: error: ‘V4L2_PIX_FMT_NV12_COL128’ undeclared here (not in a function); did you mean ‘V4L2_PIX_FMT_NV21M’?
  165 |     V4L2_PIX_FMT_NV12_COL128,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
      |     V4L2_PIX_FMT_NV21M
src/libavcodec/v4l2_request.c:166:5: error: ‘V4L2_PIX_FMT_NV12_10_COL128’ undeclared here (not in a function); did you mean ‘V4L2_PIX_FMT_NV12MT_16X16’?
  166 |     V4L2_PIX_FMT_NV12_10_COL128,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |     V4L2_PIX_FMT_NV12MT_16X16
src/libavcodec/v4l2_request.c: In function ‘v4l2_request_set_drm_descriptor’:
src/libavcodec/v4l2_request.c:189:25: error: ‘DRM_FORMAT_P030’ undeclared (first use in this function); did you mean ‘DRM_FORMAT_P010’?
  189 |         layer->format = DRM_FORMAT_P030;
      |                         ^~~~~~~~~~~~~~~
      |                         DRM_FORMAT_P010
```

seems like those are not defined...

Kernel: 5.4.28

jc-kynesim commented:
Yeah - the situation with the v4l2 and the drm header files is quite frustrating. They don't seem to be exported to userspace in the way one would like.  I've had to grab them from the linux tree.  I'm not sure what is meant to have occurred.

nkichukov commented:
Thanks @jc-kynesim 

Those seem to be defined in the kernel includes:
include/uapi/drm/drm_fourcc.h and include/uapi/linux/videodev2.h, so wonder why do not get included?

I've just defined them for the time being as per your suggestion in the libavcodec/v4l2_request.c:

```
#define V4L2_PIX_FMT_NV12_COL128 v4l2_fourcc('N', 'C', '1', '2') /* 12  Y/CbCr 4:2:0 128 pixel wide column */
#define V4L2_PIX_FMT_NV12_10_COL128 v4l2_fourcc('N', 'C', '3', '0')
                                                                /* Y/CbCr 4:2:0 10bpc, 3x10 packed as 4 bytes in
                                                                 * a 128 bytes / 96 pixel wide column */

#define DRM_FORMAT_P030         fourcc_code('P', '0', '3', '0') /* 2x2 subsampled Cr:Cb plane 10 bits per channel packed */
#define DRM_FORMAT_P010         fourcc_code('P', '0', '1', '0') /* 2x2 subsampled Cr:Cb plane 10 bits per channel */
```

and it managed to move on.

hvenev commented:
Please keep in mind the fact that you might also need to rebuild software that links against ffmpeg. The version provided at https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_4_2_2 is not ABI-compatible with upstream ffmpeg 4.2.2. I'm trying to get things to work in  https://github.com/popcornmix/ffmpeg/pull/1.
